### PR TITLE
feat: Added --ignore_feature flag

### PR DESCRIPTION
**Pull request by orenmichaely**
feat: added a flag ignore_features_dsjson_experimental

This flag ignores a feature (all values for the specified feature) in a given namespace. 

This has only been added to the dsjson parser and this functionality will be added to additional parsers when the callback functionality is added.

### CODE REVIEW

File: vowpalwabbit/core/src/parse_args.cc [hunk 1]
```diff
@@ -607,6 +608,9 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
       .add(make_option("ignore_linear", ignore_linears)
                .keep()
                .help("Ignore namespaces beginning with character <arg> for linear terms only"))
+      .add(make_option("ignore_features_dsjson_experimental", ignore_features)
+               .keep()
+               .help("Ignore specified features from namespace. To ignore a feature arg should be namespace|feature"))
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:613]:
```suggestion
               .help("Ignore specified features from namespace. To ignore a feature arg should be namespace|feature").experimental())
```

File: vowpalwabbit/core/src/parse_args.cc [hunk 2]
```diff
@@ -883,6 +887,27 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
...
   }
 
+  if (options.was_supplied("ignore_features_dsjson_experimental"))
+  {
+    for (const auto& ignored : ignore_features)
+    {
+      std::tuple<std::string, std::string> namespce_and_feature = VW::extract_ignored_feature(ignored);
+      const std::string& ns = std::get<0>(namespce_and_feature);
+      const std::string& feature_name = std::get<1>(namespce_and_feature);
+      if (!(ns.empty() || feature_name.empty()))
```
Comment (ataymano) [inline vowpalwabbit/core/src/parse_args.cc:897]: Can we ignore root namespace features? Looks like right now neither "--ignore $feature" nor "--ignore |$feature" will not enter this branch

File: vowpalwabbit/core/CMakeLists.txt [hunk 3]
```diff
@@ -403,5 +403,5 @@ endif()
 
 vw_add_test_executable(
     FOR_LIB "core"
-    SOURCES tests/cache_test.cc
-)
\ No newline at end of file
+    SOURCES tests/cache_test.cc tests/text_utils_test.cc
```
Comment (olgavrou) [inline vowpalwabbit/core/CMakeLists.txt:406]: this file can be added to the `test/unit_test/CMakeListst.txt` file

File: test/core.vwtest.json [hunk 4]
```diff
@@ -5207,7 +5207,31 @@
...
+      "ignore_feature.interactions": "train-sets/ref/ignore_feature.interactions",
+      "stdout": "train-sets/ref/ignore_feature.stdout"
+    },
+    "input_files": [
+      "train-sets/dsjson_cb_part1.json"
+    ]
+  },
+  {
+    "id": 406,
+    "desc": "test ignore multiple features",
```
Comment (olgavrou) [inline test/core.vwtest.json:5226]: per @ataymano 's point on the default namespace, might be good to add and ignore a feature without a namespace here in the tests

File: vowpalwabbit/core/CMakeLists.txt (hunk 3 above)
Comment (olgavrou) [reply vowpalwabbit/core/CMakeLists.txt:406]: never mind :)

File: vowpalwabbit/core/src/parse_args.cc (hunk 2 above)
Comment (orenmichaely) [reply vowpalwabbit/core/src/parse_args.cc:898]: good catch, fixed

File: test/core.vwtest.json (hunk 4 above)
Comment (orenmichaely) [reply test/core.vwtest.json:5226]: added :)

File: vowpalwabbit/core/src/parse_args.cc [hunk 5]
```diff
@@ -883,6 +888,27 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
     }
   }
 
+  if (options.was_supplied("ignore_features_dsjson_experimental"))
+  {
+    for (const auto& ignored : ignore_features)
+    {
+      std::tuple<std::string, std::string> namespce_and_feature = VW::extract_ignored_feature(ignored);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:895]: auto

File: vowpalwabbit/core/src/parse_args.cc [hunk 6]
```diff
@@ -883,6 +888,27 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
...
+      const std::string& ns = std::get<0>(namespce_and_feature);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:896]: const auto&

File: vowpalwabbit/core/src/parse_args.cc [hunk 7]
```diff
@@ -883,6 +888,27 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
...
+      const std::string& feature_name = std::get<1>(namespce_and_feature);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:897]: const auto&

File: vowpalwabbit/core/include/vw/core/parse_example_json.h [hunk 8]
```diff
@@ -967,14 +967,17 @@ class DefaultState : public BaseState<audit>
           *p = '_';
       }
     }
-
-    if (ctx._chain_hash) { ctx.CurrentNamespace().AddFeature(ctx.key, str, ctx._hash_func, ctx._parse_mask); }
-    else
+    const char* ns = ctx.CurrentNamespace().name;
+    if (ctx.ignore_features == nullptr || !(ctx.ignore_features->find(ns) != ctx.ignore_features->end() && ctx.ignore_features->at(ns).find(ctx.key) != ctx.ignore_features->at(ns).end()))
```
Comment (olgavrou) [inline vowpalwabbit/core/include/vw/core/parse_example_json.h:971]: nit:

replace all the negativity with some positivity? might be easier to read

```suggestion
    if (ctx.ignore_features == nullptr || (ctx.ignore_features->find(ns) == ctx.ignore_features->end() || (ctx.ignore_features ctx.ignore_features->at(ns).find(ctx.key) == ctx.ignore_features->at(ns).end()))
```

File: vowpalwabbit/core/src/text_utils.cc [hunk 9]
```diff
@@ -66,4 +66,23 @@ std::string VW::fmt_float(float f, int max_decimal_places)
...
+  if (feature_delimiter_index != std::string::npos)
+  {
+    auto ns = namespace_feature.substr(0, feature_delimiter_index);
+    // check for default namespace
+    if (ns.empty())
+    {
+      ns = " ";
+    }
+    std::get<0>(extracted_ns_and_feature) = ns;
+    std::get<1>(extracted_ns_and_feature) = namespace_feature.substr(feature_delimiter_index + 1, namespace_feature.size() - (feature_delimiter_index + 1));
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/text_utils.cc:85]:
```suggestion
    return {ns, namespace_feature.substr(feature_delimiter_index + 1, namespace_feature.size() - (feature_delimiter_index + 1))};
```

File: vowpalwabbit/core/src/text_utils.cc [hunk 10]
```diff
@@ -66,4 +66,23 @@ std::string VW::fmt_float(float f, int max_decimal_places)
...
+  }
+  return extracted_ns_and_feature;
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/text_utils.cc:87]:
```suggestion
  return {};
```

File: vowpalwabbit/core/src/text_utils.cc [hunk 11]
```diff
@@ -66,4 +66,23 @@ std::string VW::fmt_float(float f, int max_decimal_places)
   }
 
   return fmt::format("{}", f);
+}
+
+std::tuple<std::string, std::string> VW::extract_ignored_feature(const std::string& namespace_feature)
+{
+  std::tuple<std::string, std::string> extracted_ns_and_feature;
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/text_utils.cc:73]:
```suggestion
```

File: test/core.vwtest.json [hunk 12]
```diff
@@ -5207,7 +5207,44 @@
       "train-sets/dsjson_cb_part1.json",
       "train-sets/dsjson_cb_part2.json"
     ]
+  },
+  {
+    "id": 405,
+    "desc": "test ignore single feature",
+    "vw_command": "--dsjson train-sets/dsjson_cb_part1.json --cb_explore_adf --epsilon 0.2 --invert_hash ignore_feature.interactions --ignore_features_dsjson_experimental FromUrl|timeofday",
+    "diff_files": {
+      "stderr": "train-sets/ref/ignore_feature.stderr",
```
Comment (ataymano) [inline test/core.vwtest.json:5216]: do we need stdout/stderr in assert here? I believe the only invariant we are checking is the set of features in the model.
stdout/stderr is just going to be source of noise in every next outputcformat update

Comment (jackgerrits) [reply test/core.vwtest.json:5216]: It does allow us to know if any warnings are produced or error. However, now I say that we should get stdout/stderr if error code is nonzero

File: vowpalwabbit/core/include/vw/core/parse_example_json.h (hunk 8 above)
Comment (orenmichaely) [reply vowpalwabbit/core/include/vw/core/parse_example_json.h:971]: good point

File: vowpalwabbit/core/include/vw/core/text_utils.h [hunk 13]
```diff
@@ -32,6 +32,13 @@ std::string decode_inline_hex(VW::string_view arg, VW::io::logger& logger);
  */
 std::string fmt_float(float f, int max_decimal_places);
 
+/**
+ * @brief Extract namespace, feature name, and optional feature value from ignored feature string
+ *
+ * @param namespace_feature namespace|feature:feature_value. Feature value is optional and if it is supplied chain_hash is applied
+ * @return std::tuple<std::string, std::string, std::string> (namespace, feature, value = empty string)
+ */
+std::tuple<std::string, std::string> extract_ignored_feature(const std::string& namespace_feature);
```
[the remark this answers is not in the document]
Comment (orenmichaely) [reply vowpalwabbit/core/include/vw/core/text_utils.h:41]: moved to parse_args

File: vowpalwabbit/core/src/text_utils.cc (hunk 11 above)
Comment (orenmichaely) [reply vowpalwabbit/core/src/text_utils.cc:73]: moved it into parse_args

File: vowpalwabbit/core/src/text_utils.cc (hunk 10 above)
Comment (orenmichaely) [reply vowpalwabbit/core/src/text_utils.cc:87]: done

File: test/core.vwtest.json (hunk 12 above)
Comment (jackgerrits) [reply test/core.vwtest.json:5216]: I've changed my mind to agreeing with alexey :)

**Review by olgavrou — APPROVED** [review]

File: vowpalwabbit/core/include/vw/core/parse_args.h [hunk 14]
```diff
@@ -47,4 +47,12 @@ inline bool ends_with(const std::string& full_string, const std::string& ending)
...
\ No newline at end of file
+std::vector<extent_term> parse_full_name_interactions(VW::workspace& all, VW::string_view str);
+
+/**
+ * @brief Extract namespace, feature name, and optional feature value from ignored feature string
+ *
+ * @param namespace_feature namespace|feature:feature_value. Feature value is optional and if it is supplied chain_hash is applied
+ * @return std::tuple<std::string, std::string, std::string> (namespace, feature, value = empty string)
+ */
+std::tuple<std::string, std::string> extract_ignored_feature(const std::string& namespace_feature);
```
Comment (jackgerrits) [inline vowpalwabbit/core/include/vw/core/parse_args.h:58]: Please put this function in the `VW::details` namespace

File: vowpalwabbit/core/src/parse_args.cc [hunk 15]
```diff
@@ -106,6 +106,24 @@ std::string find_in_path(const std::vector<std::string>& paths, const std::strin
   return "";
 }
 
+std::tuple<std::string, std::string> extract_ignored_feature(const std::string& namespace_feature)
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:109]:
```suggestion
std::tuple<std::string, std::string> extract_ignored_feature(VW::string_view namespace_feature)
```

File: vowpalwabbit/core/src/parse_args.cc [hunk 16]
```diff
@@ -883,6 +906,27 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
     }
   }
 
+  if (options.was_supplied("ignore_features_dsjson_experimental"))
+  {
+    for (const auto& ignored : ignore_features)
+    {
+      auto namespce_and_feature = extract_ignored_feature(ignored);
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:913]: Spelling mistake:
```suggestion
      auto namespace_and_feature = extract_ignored_feature(ignored);
```

File: vowpalwabbit/core/src/parse_args.cc [hunk 17]
```diff
@@ -575,6 +593,7 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
   std::vector<std::string> full_name_interactions;
   std::vector<std::string> ignores;
   std::vector<std::string> ignore_linears;
+  std::vector<std::string> ignore_features;
```
Comment (jackgerrits) [inline vowpalwabbit/core/src/parse_args.cc:596]: Give this a more descriptive name to indicate it is only used in dsjson

Comment (jackgerrits) [reply vowpalwabbit/core/src/parse_args.cc:596]: Also, please add a comment describing its behavior

File: vowpalwabbit/core/src/parse_args.cc [hunk 18]
```diff
@@ -607,6 +629,10 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
       .add(make_option("ignore_linear", ignore_linears)
                .keep()
                .help("Ignore namespaces beginning with character <arg> for linear terms only"))
+      .add(make_option("ignore_features_dsjson_experimental", ignore_features_dsjson)
```
Comment (ataymano) [inline vowpalwabbit/core/src/parse_args.cc:632]: if there is a plan to add it to other parsers, do we still want to call it "*_dsjson_*" (and rename later)? Maybe generic name + "not supported yet" error (for other parsers) is better?

Comment (orenmichaely) [reply vowpalwabbit/core/src/parse_args.cc:632]: regardless it seems like we will have to rename it later, either by removing the "not supported yet" or the "dsjson". Once it is supported for all parsers I want the flag to be --ignore_features so all we have to do is remove the suffix. 

I dont think there is a plan to add it to all the parsers at this time, but rather refactor the parsing and add support for callback functions in the parsers. @jackgerrits will be able to elaborate more about the plan.

Comment (ataymano) [reply vowpalwabbit/core/src/parse_args.cc:632]: I didn't mean to put "not supported" to option name, but to call option "ignore_features_experimental", with runtime error / exception raised if you are trying to call --ignore_features_experimental without --dsjson. Now it seems confusing - we have 4 data formats: json, dsjson, text, flatbuffers.
1) Is this feature supported for dsjson only (but not for json)?
2) How are we going to call it if we add plain text implementation, but not flatbuffers yet?

Comment (jackgerrits) [reply vowpalwabbit/core/src/parse_args.cc:632]: Given it's experimental we can change it as much as we need, so we shouldnt worry too much about the name at the moment. When it's done and supported by all parsers then `--ignore_features` seems good. For now I think a big ugly and specific name is good

Comment (ataymano) [reply vowpalwabbit/core/src/parse_args.cc:632]: I see, makes sense.

File: test/core.vwtest.json [hunk 19]
```diff
@@ -5207,7 +5207,44 @@
       "train-sets/dsjson_cb_part1.json",
       "train-sets/dsjson_cb_part2.json"
     ]
+  },
+  {
+    "id": 405,
+    "desc": "test ignore single feature",
+    "vw_command": "--dsjson train-sets/dsjson_cb_part1.json --cb_explore_adf --epsilon 0.2 --invert_hash ignore_feature.interactions --ignore_features_dsjson_experimental FromUrl|timeofday",
```
Comment (olgavrou) [inline test/core.vwtest.json:5214]:
```suggestion
    "vw_command": "--dsjson -d train-sets/dsjson_cb_part1.json --cb_explore_adf --epsilon 0.2 --invert_hash ignore_feature.interactions --ignore_features_dsjson_experimental FromUrl|timeofday",
```

File: test/core.vwtest.json [hunk 20]
```diff
@@ -5207,7 +5207,44 @@
...
+      "stdout": "train-sets/ref/ignore_multiple_features.stdout"
+    },
+    "input_files": [
+      "train-sets/dsjson_cb_part1.json"
+    ]
+  },
+  {
+    "id": 407,
+    "desc": "test ignore feature in default namespace",
+    "vw_command": "--dsjson train-sets/dsjson_cb_default_ns.json --cb_explore_adf --epsilon 0.2 --invert_hash ignore_feature_default_ns.interactions --ignore_features_dsjson_experimental |timeofday",
```
Comment (olgavrou) [inline test/core.vwtest.json:5240]:
```suggestion
    "vw_command": "--dsjson -d train-sets/dsjson_cb_default_ns.json --cb_explore_adf --epsilon 0.2 --invert_hash ignore_feature_default_ns.interactions --ignore_features_dsjson_experimental |timeofday",
```

File: vowpalwabbit/core/src/parse_args.cc [hunk 21]
```diff
@@ -523,6 +523,25 @@ const char* are_features_compatible(VW::workspace& vw1, VW::workspace& vw2)
...
+std::tuple<std::string, std::string> extract_ignored_feature(VW::string_view namespace_feature)
+{
+  std::tuple<std::string, std::string> extracted_ns_and_feature;
+  std::string feature_delimiter = "|";
+  auto feature_delimiter_index = namespace_feature.find(feature_delimiter);
+  if (feature_delimiter_index != VW::string_view::npos)
+  {
+    auto ns = namespace_feature.substr(0, feature_delimiter_index);
+    // check for default namespace
+    if (ns.empty()) { ns = " "; }
```
Comment (ataymano) [inline vowpalwabbit/core/src/parse_args.cc:537]: I didn't know that but looks like we have counterintuitive feature collision in that kind of example:
... "feature": 1, " ": {"feature": 2} ...
@jackgerrits , is it something known?

**Review by ataymano — APPROVED** [review]

**Review by jackgerrits — APPROVED** [review]

File: vowpalwabbit/core/src/parse_args.cc (hunk 21 above)
Comment (ataymano) [reply vowpalwabbit/core/src/parse_args.cc:537]: maybe represent root namespace using some unprintable char internally?

Comment (ataymano) [reply vowpalwabbit/core/src/parse_args.cc:537]: let me create bug on that and we can discuss (or decline) there

**Pull request merged**